File: code/ex497.java
class Solution {
    public List<Integer> map = new ArrayList<>();    // Map all points inside rects to a 1D-array
    public Random rand = new Random();
    public int[][] rects;
    
    public Solution(int[][] rects) {
        this.rects = rects;
        int idx = 0;
        for (int[] rect : rects) {
            int x1 = rect[0], y1 = rect[1], x2 = rect[2], y2 = rect[3];
            idx += (x2 - x1 + 1) * (y2 - y1 + 1);
            map.add(idx);
        }
    }
    
    public int[] pick() {
        int randNum = rand.nextInt(map.get(map.size() - 1));
        
        // Find out which rect this random point is in
        int rectIdx = 0;
        for (int i = 0; i < map.size(); ++i) {
            if (map.get(i) > randNum) {
                rectIdx = i;
                break;
            }
        }
        
        // Find the actual point in this rect
        int[] targetRect = rects[rectIdx];
        int x0 = targetRect[0], y0 = targetRect[1], p0 = rectIdx == 0 ? 0 : map.get(rectIdx - 1), p = randNum - p0;
        int length = targetRect[2] - targetRect[0] + 1, height = targetRect[3] - targetRect[1] + 1;
        return new int[]{x0 + p % length, y0 + p / length};
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(rects);
 * int[] param_1 = obj.pick();
 */

File: code/ex497.cpp
class Solution {
    vector<int> points{0};
    vector<vector<int>> rects;
public:
    Solution(vector<vector<int>>& rects) {
        this->rects = rects;
        int mask = 0;
        for (auto& rect : rects) {
            mask += (rect[3] - rect[1] + 1) * (rect[2] - rect[0] + 1);
            points.push_back(mask);
        }
    }
    
    vector<int> pick() {
        int r = rand() % points.back();    // Get random number in [0, points.back())
        int idx = find(points.begin(), points.end(), *upper_bound(points.begin(), points.end(), r)) - points.begin();
        vector<int> rect = rects[idx - 1];
        int x = rect[2] - rect[0] + 1;
        return {(r - points[idx - 1]) % x + rect[0], (r - points[idx - 1]) / x + rect[1]};
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(rects);
 * vector<int> param_1 = obj->pick();
 */